flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ActiveResourceManagerFactory.java

```diff
@@ -81,4 +83,14 @@ private static Configuration createActiveResourceManagerConfiguration(Configurat
 		ClusterInformation clusterInformation,
 		@Nullable String webInterfaceUrl,
 		ResourceManagerMetricGroup resourceManagerMetricGroup) throws Exception;
+
+	protected WorkerResourceSpec createDefaultWorkerResourceSpec(Configuration configuration) {
+		final TaskExecutorProcessSpec taskExecutorProcessSpec = TaskExecutorProcessUtils
+			.newProcessSpecBuilder(configuration)
+			.withCpuCores(getDefaultCpus(configuration))
+			.build();
+		return WorkerResourceSpec.fromTaskExecutorProcessSpec(taskExecutorProcessSpec);
+	}
+
+	protected abstract CPUResource getDefaultCpus(Configuration configuration);
```

Comment:
The same applies to this method. I'm not entirely sure whether this class is the right place for this method. The `ActiveResourceManagerFactory` should not need to know how to derive these things ideally.

Not saying that we have to do it right away but I think the cleaner approach would be to pass a `WorkerResourceSpecFactory` to the `ActiveResourceManagerFactory` or even better, pass simply the default `WorkerResourceSpec` to the constructor of the `ActiveResourceManagerFactory`.

Reply:
If the deriving of cpu cores and default worker spec is too detailed for `ActiveResourceManager`, I would suggest to add an abstract method `getDefaultWorkerSpecFactory` to return different `WorkerResourceSpecFactory` for different `ActiveResourceManager` implementations.

My concern for passing `WorkerResourceSpecFactory` or `WorkerResourceSpec` to `ActiveResourceManager` is that, neither the constructor nor the `createResourceManager` method seems to be a good place to pass them in IMO.
- For the constructor, `ActiveResourceManager` is abstract and its implementations are either singleton (K8s/Yarn) or created by cluster entry point (Mesos). I think it is too detail for the cluster entry point to pass in different `WorkerResourceSpecFactory`/`WorkerResourceSpec` to the resource manager factory, not to mention breaking singleton properties on K8s/Yarn.
- For `createResourceManager`, it is called by `DefaultDispatcherResourceManagerComponentFactory`, which is not aware of the different deployments either.

Reply:
My suggestion would be to pass it to the `ActiveResourceManagerFactory` when creating it.